### ISSUE TITLE
Introduce new parsing algorithm

### ISSUE THREAD

**Issue by E1izabeth** [dbeaver/dbeaver#15524]
- Grammar model
- Grammar graph
- Finite state machine by terminals
- Grammar-based parser

**Comment by E1izabeth** [comment]
There is nothing to test yet. It's a base for future applied functionality. So, I close this issue.

**Comment by mcpiroman** [comment]
Hi, from briefly looking at the code it seems like a new parser is being built from scratch. May I ask for some terse details and motivation on why not to rely on third party (like current JSQLParser)?

In particular, for Oracle's PL/SQL, which is quite complex, I have found a very good, almost fully declarative parser https://github.com/felipebz/zpa (most grammar is specified [here](https://github.com/felipebz/zpa/blob/d1cb2fcc3bc8aa5289ee0b3c3bf3e62dfeb44bd1/zpa-core/src/main/kotlin/org/sonar/plugins/plsqlopen/api/PlSqlGrammar.kt#L222)). Can it be somehow reused? Current support for PLSQL appears to be quite minimal, maybe that can help to improve it?

Secondly, does/will DBeaver support proper semantic analysis of the parsed code? For the record I started a project to semantically analyze PL/SQL from output of the mentioned parser, maybe something like this can be done here too? It would be very helpful for working with plsql packages.
